**Instituto de Informática - UFG**

Disciplina: Inteligência Computacional

Prof. Cedric Luiz de Carvalho


Grupo (D)

---


Aluna(o): Ricardo Ataide de Oliveira Junior

---

Aluna(o): Arthur Gomes de Siqueira

---

**Engenheiro de conhecimento: Ricardo Ataide de Oliveira Junior**

###Video: https://drive.google.com/file/d/1m97o_1L84YTnAtC2uNPegawAiFG04x82/view?usp=sharing 

#Especificação do problema : Redes Neurais


A área financeira é uma área muito interessante para a aplicação de técnicas de Inteligência Artificial. Por exemplo, pode-se desenvolver aplicações que dêem suporte aos agentes financeiros na concessão ou não de crédito aos seus clientes. Pode-se também utilizar-se das técnicas de Inteligência Artificial para a identificação de atividades fraudulentas.

Utilizando a base de dados [Paysim](https://www.kaggle.com/ntnu-testimon/paysim1), projete e implemente uma rede neural para a identificação de fraudes em transações financeiras. Deve ser feita uma análise criteriosa dos dados, a fim de determinar os hiperparâmetros da rede. Devem ser gerados gráficos que possam dar uma visão geral da base de dados. Os resultados da qualidade do modelo implementado devem estar explícitos no relatório técnico. Devem ser realizados diversos testes e a qualidade do modelo medida. 

Após implementada a solução, deve-se comparar os resultados obtidos com aqueles obtidos no trabalho anterior. Analise tecnicamente qual das duas técnicas é melhor para resolver o problema em questão.



## O trabalho

O trabalho deve ser feito em grupos, conforme definido. Cada grupo deverá entregar apenas um produto final para ser avaliado pelo professor. Apesar do trabalho ser feito em grupos, a avaliação de cada aluno será feita individualmente.


#Proposta de solução para o problema

In [ ]:
import pandas as pd
import plotly.express as px

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
df = pd.read_csv('/content/drive/MyDrive/Raiz/Faculdade/2021.1/IC/Trabalhos/Atividade 6 - Arvores de decisao/base_corrigida.csv')
df.head(10)

Unnamed: 0  step      type  ...  newbalanceDest isFraud  isFlaggedFraud
0           0     1   PAYMENT  ...            0.00       0               0
1           1     1   PAYMENT  ...            0.00       0               0
2           2     1  TRANSFER  ...            0.00       1               0
3           3     1  CASH_OUT  ...            0.00       1               0
4           4     1   PAYMENT  ...            0.00       0               0
5           5     1   PAYMENT  ...            0.00       0               0
6           6     1   PAYMENT  ...            0.00       0               0
7           7     1   PAYMENT  ...            0.00       0               0
8           8     1   PAYMENT  ...            0.00       0               0
9           9     1     DEBIT  ...        40348.79       0               0

[10 rows x 12 columns]

In [ ]:
#Ajustando a coluna 'type' para possibilitar a utilização de RNAS

colunas = df.type.unique()
df_correcao = pd.DataFrame(columns = colunas)
df_correcao

Empty DataFrame
Columns: [PAYMENT, TRANSFER, CASH_OUT, DEBIT, CASH_IN]
Index: []

dict_values([1, 0, 0, 0, 0])

In [ ]:
#Populando o df
colunas = df.type.unique()
def parametriza(df,type):
#Funcção responsavel por transformar os dados do tipo de movimentação em binários
#Recebera o campo que deve ser populado no df, no caso o type indica isso
  if (type == 'PAYMENT'):
    if df == colunas[0]:
      #Retorna 1 se for payment retorna 0 se nao for, o mesmo para os demais campos com seus respectivos types
      return 1
    else: 
      return 0

  elif (type == 'TRANSFER'):
    if df == colunas[1]:
      return 1
    else: 
      return 0
  
  elif (type == 'CASH_OUT'):
    if df == colunas[2]:
      return 1
    else: 
      return 0
    
  elif (type == 'DEBIT'):
    if df == colunas[3]:
      return 1
    else: 
      return 0

  elif (type == 'CASH_IN'):
    if df == colunas[4]:
      return 1
    else: 
      return 0

  print(len(df_colunas))
print('Resultado: ',len(df) - len(df_correcao))

Resultado:  3862200


In [ ]:
#Criando o 'df2' para que esse possua os dados já 'parametrizados'
df2 = df
df2['PAYMENT'] = (df['type'].apply(lambda x: parametriza(x,'PAYMENT')))
df2['TRANSFER'] = (df['type'].apply(lambda x: parametriza(x,'TRANSFER')))
df2['CASH_OUT'] = (df['type'].apply(lambda x: parametriza(x,'CASH_OUT')))
df2['DEBIT'] = (df['type'].apply(lambda x: parametriza(x,'DEBIT')))
df2['CASH_IN'] = (df['type'].apply(lambda x: parametriza(x,'CASH_IN')))
df2.head(10)



Unnamed: 0  step      type    amount  ... TRANSFER  CASH_OUT  DEBIT CASH_IN
0           0     1   PAYMENT   9839.64  ...        0         0      0       0
1           1     1   PAYMENT   1864.28  ...        0         0      0       0
2           2     1  TRANSFER    181.00  ...        1         0      0       0
3           3     1  CASH_OUT    181.00  ...        0         1      0       0
4           4     1   PAYMENT  11668.14  ...        0         0      0       0
5           5     1   PAYMENT   7817.71  ...        0         0      0       0
6           6     1   PAYMENT   7107.77  ...        0         0      0       0
7           7     1   PAYMENT   7861.64  ...        0         0      0       0
8           8     1   PAYMENT   4024.36  ...        0         0      0       0
9           9     1     DEBIT   5337.77  ...        0         0      1       0

[10 rows x 17 columns]

In [ ]:
#Realizando mais alguns ajustes finos no df2, removendo colunas não mais necessárias
df2 = df2.drop(columns=['type', 'Unnamed: 0'])
df2

step      amount     nameOrig  ...  CASH_OUT  DEBIT CASH_IN
0           1     9839.64  C1231006815  ...         0      0       0
1           1     1864.28  C1666544295  ...         0      0       0
2           1      181.00  C1305486145  ...         0      0       0
3           1      181.00   C840083671  ...         1      0       0
4           1    11668.14  C2048537720  ...         0      0       0
...       ...         ...          ...  ...       ...    ...     ...
3862195   743   339682.13   C786484425  ...         1      0       0
3862196   743  6311409.28  C1529008245  ...         0      0       0
3862197   743  6311409.28  C1162922333  ...         1      0       0
3862198   743   850002.52  C1685995037  ...         0      0       0
3862199   743   850002.52  C1280323807  ...         1      0       0

[3862200 rows x 15 columns]

In [ ]:
#Reordenando a coluna 'isFraud" por facilidade no momento de definir os parametros de INPUT E OUTPUT
isFraud = df2.isFraud
df2 = df2.drop(columns=['isFraud'])
df2 = pd.concat([df2,isFraud],axis=1) 

In [ ]:
df2.head()

step    amount     nameOrig  ...  DEBIT  CASH_IN isFraud
0     1   9839.64  C1231006815  ...      0        0       0
1     1   1864.28  C1666544295  ...      0        0       0
2     1    181.00  C1305486145  ...      0        0       1
3     1    181.00   C840083671  ...      0        0       1
4     1  11668.14  C2048537720  ...      0        0       0

[5 rows x 15 columns]

In [ ]:
#Criando duas versões do dataframe, uma somente com dados binários e outra com os dados de remetente e destinatrio
df2.to_csv('/content/drive/MyDrive/Raiz/Faculdade/2021.1/IC/Trabalhos/Atividade 7 - RNA/df_param_comp.csv')
df2 = df2.drop(columns=['nameOrig','nameDest'])
df2.to_csv('/content/drive/MyDrive/Raiz/Faculdade/2021.1/IC/Trabalhos/Atividade 7 - RNA/df_param_simp.csv')
#Salvando os dados em csvs para facilitar novas analises

In [ ]:
#Definindo a rede de teste com apenas 100 mil registros
df_full = pd.read_csv('/content/drive/MyDrive/Raiz/Faculdade/2021.1/IC/Trabalhos/Atividade 7 - RNA/df_param_simp.csv')
df_teste_simp = df_full.head(100000)
print(len(df_teste_com))
print(len(df_teste_simp))

100000
100000



#Inicio da implementação

Código baseado no modelo de implementação apresentado pelo professor [Cedric](https://turing.inf.ufg.br/mod/resource/view.php?id=13435), o qual por sua vez foi baseado no trabalho de [Sagar Howal](https://towardsdatascience.com/neural-networks-with-google-colaboratory-artificial-intelligence-getting-started-713b5eb07f14) 

## Solução utilizando toda a base de dados (aproximadamente 3.6*10⁶ registros)

In [ ]:
df_full = pd.read_csv('/content/drive/MyDrive/Raiz/Faculdade/2021.1/IC/Trabalhos/Atividade 7 - RNA/df_param_simp.csv')
len(df_teste_comp)

3862200

In [ ]:
df_full.head()

Unnamed: 0  step    amount  oldbalanceOrg  ...  CASH_OUT  DEBIT  CASH_IN  isFraud
0           0     1   9839.64       170136.0  ...         0      0        0        0
1           1     1   1864.28        21249.0  ...         0      0        0        0
2           2     1    181.00          181.0  ...         0      0        0        1
3           3     1    181.00          181.0  ...         1      0        0        1
4           4     1  11668.14        41554.0  ...         0      0        0        0

[5 rows x 14 columns]

In [ ]:
# #Definindo as variáveis dependentes e independentes
x = df_full.iloc[:,1:13] #Caso sejam removidos dados nao numericos
y = df_full.iloc[:,13]

In [ ]:
import numpy as np
#Convertendo para array os dados de treinamento
x = np.asarray(x).astype('float32')
y = np.asarray(y).astype('float32')
len(df) - len(x)

0

In [ ]:
#Separando os dados em dados de treinamento e teste
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.35, random_state = 42)

In [ ]:
# Importando as bibliotecas do pacote Keras paratrabalhar com RNAs
import keras
from keras.models import Sequential
from keras.layers import Dense

In [ ]:
#Inicialização da rede
classifier = Sequential()
classifier

In [ ]:
len(df_full[df_full['isFraud'] == 1])

8184

### Inicializando a RNA

In [ ]:
#Adicionando a primeira Camada da rede
classifier.add(Dense(units = 7, kernel_initializer='uniform', activation = 'relu', input_dim = 12))

#Adicionando a segunda camada da rede
classifier.add(Dense(units = 7, kernel_initializer='uniform', activation = 'relu'))

#Adicionando a camada de saída
classifier.add(Dense(units = 1, kernel_initializer='uniform', activation='sigmoid'))

# Compilando a RNA
classifier.compile(optimizer = 'adam', loss='binary_crossentropy', metrics=['accuracy'])



In [ ]:
#Adequando a RNA ao conjunto de exemplos
classifier.fit(x_train, y_train, batch_size = 12, epochs = 5)
#batch_size: o número de entradas processadas simultaneamente
#epoch indica a passagem completa sobre todo o dataset

Epoch 1/5
209203/209203 [==============================] - 273s 1ms/step - loss: 0.2515 - accuracy: 0.9978
Epoch 2/5
209203/209203 [==============================] - 273s 1ms/step - loss: 0.2552 - accuracy: 0.9978
Epoch 3/5
209203/209203 [==============================] - 271s 1ms/step - loss: 0.5791 - accuracy: 0.9978
Epoch 4/5
209203/209203 [==============================] - 270s 1ms/step - loss: 0.1888 - accuracy: 0.9978
Epoch 5/5
209203/209203 [==============================] - 271s 1ms/step - loss: 0.0152 - accuracy: 0.9979


In [ ]:
# Fazendo as previsões sobre o conjunto de testes
y_pred = classifier.predict(x_test)
y_pred = (y_pred > 0.5)

# Construindo a Matriz de Confusão
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred) 

In [ ]:
cm = confusion_matrix(y_test, y_pred) 
cm

array([[1348917,       0],
       [   2853,       0]])

## Utilizando somente funções 'relu'

In [ ]:
#Inicialização da rede
classifier = Sequential()

#Adicionando a primeira Camada da rede
classifier.add(Dense(units = 7, kernel_initializer='uniform', activation = 'relu', input_dim = 12))

#Adicionando a segunda camada da rede
classifier.add(Dense(units = 7, kernel_initializer='uniform', activation = 'relu'))

#Adicionando a camada de saída
classifier.add(Dense(units = 1, kernel_initializer='uniform', activation='relu'))

# Compilando a RNA
classifier.compile(optimizer = 'adam', loss='binary_crossentropy', metrics=['accuracy'])

classifier.fit(x_train, y_train, batch_size = 12, epochs = 5)

# Fazendo as previsões sobre o conjunto de testes
y_pred = classifier.predict(x_test)
y_pred = (y_pred > 0.5)

# Construindo a Matriz de Confusão
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred) 
cm

Epoch 1/5
257480/257480 [==============================] - 317s 1ms/step - loss: 0.0332 - accuracy: 0.9978
Epoch 2/5
257480/257480 [==============================] - 319s 1ms/step - loss: 0.0329 - accuracy: 0.9979
Epoch 3/5
257480/257480 [==============================] - 319s 1ms/step - loss: 0.0329 - accuracy: 0.9979
Epoch 4/5
257480/257480 [==============================] - 319s 1ms/step - loss: 0.0329 - accuracy: 0.9979
Epoch 5/5
257480/257480 [==============================] - 320s 1ms/step - loss: 0.0329 - accuracy: 0.9979


array([[770838,      0],
       [  1602,      0]])

## Utilizando somente funções 'Sigmoid'

In [ ]:
#Inicialização da rede
classifier = Sequential()

#Adicionando a primeira Camada da rede
classifier.add(Dense(units = 7, kernel_initializer='uniform', activation = 'sigmoid', input_dim = 12))

#Adicionando a segunda camada da rede
classifier.add(Dense(units = 7, kernel_initializer='uniform', activation = 'sigmoid'))

#Adicionando a camada de saída
classifier.add(Dense(units = 1, kernel_initializer='uniform', activation='sigmoid'))

# Compilando a RNA
classifier.compile(optimizer = 'adam', loss='binary_crossentropy', metrics=['accuracy'])

classifier.fit(x_train, y_train, batch_size = 12, epochs = 5)

# Fazendo as previsões sobre o conjunto de testes
y_pred = classifier.predict(x_test)
y_pred = (y_pred > 0.5)

# Construindo a Matriz de Confusão
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred) 
cm

Epoch 1/5
257480/257480 [==============================] - 327s 1ms/step - loss: 0.0158 - accuracy: 0.9979
Epoch 2/5
257480/257480 [==============================] - 324s 1ms/step - loss: 0.0153 - accuracy: 0.9979
Epoch 3/5
257480/257480 [==============================] - 313s 1ms/step - loss: 0.0153 - accuracy: 0.9979
Epoch 4/5
257480/257480 [==============================] - 309s 1ms/step - loss: 0.0153 - accuracy: 0.9979
Epoch 5/5
257480/257480 [==============================] - 313s 1ms/step - loss: 0.0153 - accuracy: 0.9979


array([[770838,      0],
       [  1602,      0]])

## Teste com 10 épocas

In [ ]:
#Inicialização da rede
classifier = Sequential()

#Adicionando a primeira Camada da rede
classifier.add(Dense(units = 7, kernel_initializer='uniform', activation = 'relu', input_dim = 12))

#Adicionando a segunda camada da rede
classifier.add(Dense(units = 7, kernel_initializer='uniform', activation = 'relu'))

#Adicionando a camada de saída
classifier.add(Dense(units = 1, kernel_initializer='uniform', activation='relu'))

# Compilando a RNA
classifier.compile(optimizer = 'adam', loss='binary_crossentropy', metrics=['accuracy'])

classifier.fit(x_train, y_train, batch_size = 12, epochs = 10)

# Fazendo as previsões sobre o conjunto de testes
y_pred = classifier.predict(x_test)
y_pred = (y_pred > 0.5)

# Construindo a Matriz de Confusão
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred) 
cm

Epoch 1/10
209203/209203 [==============================] - 278s 1ms/step - loss: 0.0336 - accuracy: 0.9978
Epoch 2/10
209203/209203 [==============================] - 277s 1ms/step - loss: 0.0328 - accuracy: 0.9979
Epoch 3/10
209203/209203 [==============================] - 276s 1ms/step - loss: 0.0328 - accuracy: 0.9979
Epoch 4/10
209203/209203 [==============================] - 278s 1ms/step - loss: 0.0328 - accuracy: 0.9979
Epoch 5/10
209203/209203 [==============================] - 277s 1ms/step - loss: 0.0328 - accuracy: 0.9979
Epoch 6/10
209203/209203 [==============================] - 277s 1ms/step - loss: 0.0328 - accuracy: 0.9979
Epoch 7/10
209203/209203 [==============================] - 276s 1ms/step - loss: 0.0328 - accuracy: 0.9979
Epoch 8/10
209203/209203 [==============================] - 276s 1ms/step - loss: 0.0328 - accuracy: 0.9979
Epoch 9/10
209203/209203 [==============================] - 266s 1ms/step - loss: 0.0328 - accuracy: 0.9979
Epoch 10/10
209203/209203 [=

array([[1348917,       0],
       [   2853,       0]])

## Comparando com Árvores de decisão (trabalho anterior)

Na implementação de árvores de decisão para solução deste mesmo problema, também foi alcançada uma elevada acurácia preditiva (99%), mas então, qual seria a melhor abordagem para o problema?

A resposta para esta pergunta é complexa. 

Primeiro vamos fazer um comparativo entre as técnicas:

Árvores de Decisão 
--- 
* Definição de modelo simplificada, as técnicas de auto-analise das árvores de decisão permitem um modelo simples de se definir 
* Elevado custo de memória, uma arvore de decisão pode chegar a ter uma estrutura muito grande para um dado problema 
* Fácil compreensão e análise 
---
RNA
---
* Definição complexa, devido as inúmeras arquiteturas possíveis, é dificil decidir qual será a mais adequada para a solução do problema
* Um problema que seria necessário uma grande árvore, é resolvido por poucos neurônios
*  Modelo de dificil compreensão e complexo de ser análisado.
---

Apesar das diferenças entre ambos os modelos é notório que ambos alcançaram o objetivo de resolver o problema proposto com soluções de qualidades equiparáveis, no entanto devido ao rápido crescimento na estrutura de uma árvore de decisão a mesma é mais recomendada para problemas de menor complexidade, enquanto problemas mais complexos, podem ser resolvidos por RNAs, mesmo que com certa complexidade.



## Análise da solução

A solução implementada apresentou-se eficaz, chegando a uma acurácia de 99.79%, em uma rede relativamente simples, com apenas 11 neurônios. 

Como já dito anteriormente a base utilizada ainda careece de variaveis relevantes que podem ser consideradas no 'mundo real' e possuem elevada relevância para a solução do problema, apesar disso, olhando somente para 'mundo' que pode ser modelado, a solução pode ser tido como boa, apensar de, ao ser comparada com o mundo real, padecer de informações.

Um trabalho mais completo abordadndo o tema pode ser encontrado em:


https://thesai.org/Downloads/Volume10No9/Paper_43-Fraud_Detection_using_Machine_Learning_in_E_Commerce.pdf


